File: Sorting/Selection-Sort/SelectionSort.py
"""
Selection sort using python
Time Complexity - O(n^2)
Space Complexity - O(1)
"""


def selection_sort(arr):
    current_index = 0
    while current_index < len(arr) - 1:
        smallest_index = current_index
        for i in range(current_index + 1, len(arr)):
            if arr[smallest_index] > arr[i]:
                smallest_index = i
        [arr[current_index], arr[smallest_index]] = [
            arr[smallest_index], arr[current_index]]
        current_index += 1

    return arr


if __name__ == "__main__":
    print(selection_sort([2, 3, 1, 8, 5, 4, 11, 9, 19, 12]))

File: Sorting/Selection-Sort/SelectionSort.ts
// Selection sort in TypeScript
// time complexity = O(n^2)
// space complexity  = O(1)

export function SelectionSort(arr:number[]):number[]{
    
    let currentIndex = 0;
    while(currentIndex < arr.length -1){
        let smallestIndex = currentIndex;
        for(let i = currentIndex+1; i < arr.length; i++){
            if(arr[smallestIndex]>arr[i]){
                smallestIndex  = i;

            }
        }
        [arr[smallestIndex],arr[currentIndex]]= [arr[currentIndex],arr[smallestIndex]];
        currentIndex +=1;
    }
    return arr;
}

console.log(SelectionSort([4,3,21,1,4,11]));